### PR TITLE
mgmt: mcumgr: Fix Bluetooth transport issues

### LINKED ISSUE #51653: mgmt: mcumgr: bt: issue with queued packets when device is busy
**Describe the bug**
This issue is similar to https://github.com/zephyrproject-rtos/zephyr/issues/51594 but is slightly different, the same transport is used (bluetooth) but it requires a specific flow to trigger this:
1. Remote device connects
2. Device issues a command that takes some time to run, e.g. img_mgmt slot erase, where **the bluetooth stack/radio operation is not impeded**, this means that this issue will not arise if the secondary slot is in internal flash as the stack cannot run with an ongoing flash erase, however it can run when erasing an external QSPI flash memory
3. Remote device disconnects
4. Remote device connects
5. Remote device queues up an operation e.g. img_mgmt write data
6. Remote device disconnects
At this point, there is an ongoing flash erase, and a queued command (which has a reference count > 0) which means the device cannot restart advertising, it will then process the flash write command after the erase has completed.

**Expected behavior**
The pending operation should not be queued if there is an ongoing operation

**Impact**
Showstopper, device communications are dead until the unit is rebooted

**Environment (please complete the following information):**
 - Commit SHA or Version used: c707f1ed28a83d9a0715ca08743d0e2112a3fa95

### LINKED ISSUE #51714: Bluetooth: Application with buffer that cannot unref it in disconnect handler leads to advertising issues
**Describe the bug**
This issue has been found with a peripheral zephyr device which supports 1 connection, the application works by receiving BT data and then putting it into a fifo to be processed by a dedicated (non-system) workqueue which will send responses to it and unref the connection, an issue arises however in this application if the remote device disconnects prior to the task having finished, the application attempts to restart advertising using the system workqueue from the disconnection handler. The problem is that because the other task contains references to the connection, restarting advertising fails with error -12. Therefore the application is unable to restart advertising and then is in a dead state. Upping the connection count to 2 does work around this issue, but is not an acceptable solution due to the additional RAM usage (which is essentially wasted).
Some operations can also take longer than others, e.g. a image erase command can take upwards of 5 seconds to process.

**To Reproduce**
mcumgr has this issue

**Expected behavior**
To be able to restart advertising, from the system workqueue, triggered in the disconnect handler

**Impact**
Showstopper

**Environment (please complete the following information):**
 - Commit SHA or Version used: This is apparent in an nRF connect SDK sample application (samples/matter/lock) using the softdevice, but should be reproducible with any application that uses bluetooth, I did not attempt seeing if this issue was present in the zephyr controller but would imagine it would be given that it is caused by the limit of connection references/handles being reached

### LINKED ISSUE #52022: RFC: API Change: mgmt: mcumgr: transport: Add query valid check function
## Problem Description:

This is a stable API change request for MCUmgr to add a new function pointer to transports to be used in a callback to check if pending queries are still valid or invalid (and, if invalid, to purge them from the queue). The requirement for this is with connection-orientated transports, e.g. Bluetooth - if a connection is dropped then it does not make sense to continue running MCUmgr commands that it queued up.

## Proposed Change (detailed):

This change is to add a new function pointer to the `smp_transport_init()` function which takes takes a netbuf pointer and void pointer (the void pointer is supplied when the `smp_rx_remove_invalid()` function is called) and returns false if a netbuf is no longer valid and should be purged, or true if it is still valid.

PR: https://github.com/zephyrproject-rtos/zephyr/pull/51757

For out-of-tree zephyr application code implementing MCUmgr transports, this requires either implementing the function (for transports that would need to support it) or setting the function to NULL in the `smp_transport_init()` call, in which case the transport will behave as it does now. If a transport implements this functionality, it will also need to make a call to `smp_rx_remove_invalid()` when the connection is no longer valid to start the cleaning up process

### PR DESCRIPTION
**Pull request by nordicjm**
This fixes issues with the Bluetooth SMP transport whereby deadlocks could arise from connection references being held in long-lasting mcumgr command processing functions.

Fixes #51653
Fixes #51714

### CODE REVIEW

File: include/zephyr/mgmt/mcumgr/smp.h
Comment (nordicjm) [inline include/zephyr/mgmt/mcumgr/smp.h:87]: Find x (10 marks)

File: subsys/mgmt/mcumgr/transport/smp_bt.c
Comment (nordicjm) [inline subsys/mgmt/mcumgr/transport/smp_bt.c:495]: Junk this if it is no longer needed

File: subsys/mgmt/mcumgr/transport/smp_bt.c
Comment (jori-nordic) [inline subsys/mgmt/mcumgr/transport/smp_bt.c:599]: If I understand correctly, this is checking that the smp buffer belongs to the conn obj passed in `arg`, right?
Why do we need the `cpd->id == 0` part?

edit: I see this block in `connected`
```c
			if (cpd->id == 0) {
				++cpd->id;
				++next_id;
			}
```
does that mean that `id==0` has a special meaning?

**Review by jori-nordic — CHANGES_REQUESTED** [review]

File: subsys/mgmt/mcumgr/transport/smp_bt.c
Comment (de-nordic) [reply subsys/mgmt/mcumgr/transport/smp_bt.c:524]: What are the chances that `connected` gets preempted and other `connected` instance gets called?
We can use CPU ticks here or something like that, there is very low chance that a connection will be held long enough to wrap around and get the same ticks for two different connections, and even in code executed in parallel, unless two CPUs are used, ticks should differ.
We can probably use only the lower 32bits of  `k_uptime_ticks(void)` safely.

**Comment by jori-nordic** [comment]
@de-nordic https://github.com/zephyrproject-rtos/zephyr/pull/51757#discussion_r1014223549

Yeah, I will bring this up with the other bluetooth maintainers, as I've seen a bunch of cases where we want to do exactly that, check that the conn object is still pointing to the connection (session?) we thought it was.
Maybe having the cpu ticks or uptime would be enough, yeah. It would also double-up as a session-length counter (until it wraps, that is). 
I'll see what I can do to have something like that as an official API sometime in the near future.

File: subsys/mgmt/mcumgr/transport/smp_bt.c
Comment (nordicjm) [reply subsys/mgmt/mcumgr/transport/smp_bt.c:625]: *Count of 0, the default will be 0 and maximum will be 1 (per connection). The idea being that for each connection, you can send 1 message at a time and have to wait for the sent callback semaphore to be given

File: subsys/mgmt/mcumgr/transport/smp_bt.c
Comment (nordicjm) [reply subsys/mgmt/mcumgr/transport/smp_bt.c:524]: > I have some trouble following why this ID is needed, could you elaborate?
The ID is needed because a connection can be established, a mcumgr command can be sent, then that device can disconnect and a new device can connect - if this is a long running command it can take seconds to run, at which point without the ID check it will send the response to the new device which is invalid operation. In terms of a race condition, this is executed in the system workqueue, which runs to completion, and would need to exhaust 255 entries to loop around (which, assuming someone was connecting and disconnecting continuously, assuming it would take 30ms to do a cycle, that would take 7.65 seconds)

> What are the chances that connected gets preempted and other connected instance gets called
0 as far as I'm aware as it's called in the system workqueue which is blocked until it has completed (co-operative thread)

File: subsys/mgmt/mcumgr/transport/smp_bt.c
Comment (nordicjm) [reply subsys/mgmt/mcumgr/transport/smp_bt.c:599]: ID = 0 means unused, ID != 0 means used

File: subsys/mgmt/mcumgr/smp.c
Comment (nordicjm) [reply subsys/mgmt/mcumgr/smp.c:176]: Clear all is for future use, I've left it here so to avoid going through a stable API change. The usage would be for something like the UART transport, if a user wants to disable mcumgr then all pending queries can be cleared to use the UART for some other purpose

File: subsys/mgmt/mcumgr/transport/smp_bt.c
Comment (jori-nordic) [reply subsys/mgmt/mcumgr/transport/smp_bt.c:625]: my bad, I read too fast and didn't see the index for the semaphore

**Comment by de-nordic** [comment]
> @de-nordic [#51757 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/51757#discussion_r1014223549)
> 
> Yeah, I will bring this up with the other bluetooth maintainers, as I've seen a bunch of cases where we want to do exactly that, check that the conn object is still pointing to the connection (session?) we thought it was. Maybe having the cpu ticks or uptime would be enough, yeah. It would also double-up as a session-length counter (until it wraps, that is). I'll see what I can do to have something like that as an official API sometime in the near future.

Having some unique identifier passed from BLE susbystem would be nice, because that would free us, and other subsystems, from figuring out to add additional layer of identification.
Hashes on objects does not make sense, due to statistical reasons, and random values would require recording values in use.  Value like ticks or time since system startup or first connection are increasing automatically (are monotonic, as much as type size allows before they overlap) giving unique identifier each time when picked as long as precision allows to obtain different values for closest possible connections, and as long as they have range that is able to hold value longer than time the connection information is held by a system.

**Review by de-nordic — CHANGES_REQUESTED** [review]
Please release notes entry for MCUmgr describing addition, and to fixed bugs section.

**Review by jori-nordic — DISMISSED** [review]

**Comment by de-nordic** [comment]
Question: why aren't we allocating the id in `smp_bt_chr_write` after this:
https://github.com/zephyrproject-rtos/zephyr/blob/4bae4f426ff4c5e4b84f29d8150789db1c7c569b/subsys/mgmt/mcumgr/transport/smp_bt.c#L262-L264
and before
https://github.com/zephyrproject-rtos/zephyr/blob/4bae4f426ff4c5e4b84f29d8150789db1c7c569b/subsys/mgmt/mcumgr/transport/smp_bt.c#L233

In case when reassembly is enabled this basically can be done by having static counter within function that will be incremented every time function is invoked.

Non re-assembly case is a little bit trickier because there may be multiple buffers per request, but this can probably be resolved my doing conn to id map.

File: subsys/mgmt/mcumgr/transport/smp_bt.c
Comment (nordicjm) [reply subsys/mgmt/mcumgr/transport/smp_bt.c]: Fixed

**Comment by nordicjm** [comment]
> Question: why aren't we allocating the id in smp_bt_chr_write after this:

The way it works here is much simpler and works for all queued packets, i.e. a device can queue up 5 commands and disconnect, it does not need different IDs for each command, they should be the same.

File: subsys/mgmt/mcumgr/transport/smp_bt.c
Comment (nordicjm) [reply subsys/mgmt/mcumgr/transport/smp_bt.c:437]: A command can be sent over 3 packets, if the device disconnects after receiving the first packet then cpd->id will be updated to 0, 0 meaning invalid, and this check will then hit. 0 is reserved for "invalid ID", it specifically means that the connection is not valid.

**Comment by nordicjm** [comment]
Reverted back to an incrementing counter as I don't see the value in tick count, it adds nothing useful and still needs the same != 0 check. Have allowed cpd to be null instead of asserting in the get function and added null checks and LOG_ERR() output on it being null, moved ID generation to alloc function.

**Review by jori-nordic — APPROVED** [review]
Advertising is getting some attention recently in the host, so I'll make a task to fix this behavior properly (ie restart advertising after the only conn obj is freed).

**Review by carlescufi — CHANGES_REQUESTED** [review]
Please add an entry in the release notes

**Comment by nordicjm** [comment]
Release notes on https://github.com/zephyrproject-rtos/zephyr/pull/52098

**Review by de-nordic — APPROVED** [review]
The new callbacks look OK.
I do not like that we have to generate own unique IDs for connections, as it should be done by BT, but it is only workaround we can afford now.

**Review by carlescufi — APPROVED** [review]

**Comment by cfriedt** [comment]
@nordicjm - in order to backport this change, the `backport v2.7-branch` label should be added to this PR. Is there a reason that wasn't done? I should mention, that even if an automated backport fails, you can still modify it (it gets a special backport branch).

**Comment by nordicjm** [comment]
@cfriedt I didn't do it from this because the code in 2.7 is so vastly outdated compared to main (e.g. different filenames and paths, different functions, different code) that none of it applies and it was just easier to manually copy and paste the bits into the 2.7 tree

**Pull request merged**
